src/fromager/constraints.py

```diff
@@ -0,0 +1,54 @@
+import copy
+import logging
+import pathlib
+from io import TextIOWrapper
+
+from packaging.requirements import Requirement
+
+logger = logging.getLogger(__name__)
+
+
+class Constraints:
+    def __init__(self, data: dict[str, Requirement]):
+        self._data = data
+
+    def get_new_requirement(self, req: Requirement):
+        constraint = self._data.get(req.name)
+        new_req = copy.deepcopy(req)
+
+        if not constraint:
+            return new_req, None
+
+        # only allow "==" in constraints for now
+        if len(constraint.specifier) != 1 or not str(constraint.specifier).startswith(
+            "=="
+        ):
+            logger.debug(
+                f"{constraint} is not allowed. Only '<package>==<version>' constraints are allowed"
+            )
+            return new_req, None
+
+        new_req.specifier = constraint.specifier
+        return new_req, constraint
+
+
+def _parse(content: TextIOWrapper) -> Constraints:
+    constraints = {}
+    for line in content:
+        req = Requirement(line.strip())
+        constraints[req.name] = req
+    return Constraints(constraints)
+
+
+def load(filename: pathlib.Path | None) -> Constraints:
+    if not filename:
+        return Constraints({})
+    filepath = pathlib.Path(filename)
+    if not filepath.exists():
+        logger.debug(
```

Comment:
A missing file is almost certainly a user error that should not be ignored.

Reply:
I copied this part from `settings.py`. Will make a note to change it there as well in a separate patch

Reply:
Ah, in the case of the settings file there's a default and providing the file is optional. In this case, the user will be explicitly giving us a file and there is no default, so if they point us to something that doesn't exist it's more clearly a mistake.

Reply:
Oh got it